File: java/main/com/latencybusters/lbmct/Tmr.java
package com.latencybusters.lbmct;
/*
 * See https://github.com/UltraMessaging/lbmct for code and documentation.
 *
 * Copyright (c) 2018-2019 Informatica Corporation. All Rights Reserved.
 * Permission is granted to licensees to use or alter this software for
 * any purpose, including commercial applications, according to the terms
 * laid out in the Software License Agreement.
 -
 - This receiver code example is provided by Informatica for educational
 - and evaluation purposes only.
 -
 - THE SOFTWARE IS PROVIDED "AS IS" AND INFORMATICA DISCLAIMS ALL WARRANTIES
 - EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION, ANY IMPLIED WARRANTIES OF
 - NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
 - INFORMATICA DOES NOT WARRANT THAT USE OF THE SOFTWARE WILL BE UNINTERRUPTED
 - OR ERROR-FREE.  INFORMATICA SHALL NOT, UNDER ANY CIRCUMSTANCES, BE
 - LIABLE TO LICENSEE FOR LOST PROFITS, CONSEQUENTIAL, INCIDENTAL, SPECIAL OR
 - INDIRECT DAMAGES ARISING OUT OF OR RELATED TO THIS AGREEMENT OR THE
 - TRANSACTIONS CONTEMPLATED HEREUNDER, EVEN IF INFORMATICA HAS BEEN APPRISED
 - OF THE LIKELIHOOD OF SUCH DAMAGES.
 */

import java.util.concurrent.*;
import com.latencybusters.lbm.*;

/**
 * Wrapper for UM timers with semantics that are more deterministic.
 * When a pending UM timer is canceled, it is possible that the cancel
 * comes a little too late and the timer callback is already being
 * invoked.  In that case, the timer cancel method can return but the
 * callback is still being executed.
 * <p>
 * This wrapper performs a handshake with the underlying UM context
 * to guarantee that, when the cancel method returns, the callback
 * either won't be invoked, or has completed.
 */
class Tmr {
  enum States {
    IDLE, TIMING, CANCEL_PENDING
  }

  private LBMContext ctx;
  private States tmrState;
  private int id;
  private Semaphore wakeupSem;
  private TmrCallback cb;
  private Object cbArg;
  private LBMTimer umTimer;
  private LBMTimer umTimerRecycler;  // re-use old UM timer object.
  private LBMTimer umCanceler;
  private LBMTimer umCancelerRecycler;  // re-use old UM timer object.
  private TmrExpireCallback expireCb;
  private TmrCancelerCallback cancelerCb;

  /**
   * Creates a UM timer wrapper with the given UM context.
   * @param ctx UM context to be used for the underlying UM timers.
   */
  Tmr(LBMContext ctx) {
    this.ctx = ctx;
    tmrState = States.IDLE;
    id = 0;
    wakeupSem = new Semaphore(0);
    cb = null;
    cbArg = null;
    umTimer = null;
    umTimerRecycler = null;
    umCanceler = null;
    umCancelerRecycler = null;
    expireCb = new TmrExpireCallback(this, ctx);
    cancelerCb = new TmrCancelerCallback(this, ctx);
  }

  // Public API to see the ID of the current (or most recent) timer.
  // THREAD: user
  int getId() {
    return id;
  }

  // THREAD: user
  // Public API to schedule a timer.  May be called from any thread.
  int schedule(TmrCallback cb, Object cbArg, int delay) throws Exception {
    int rtnId;
    synchronized (this) {
      if (tmrState != States.IDLE) {
        throw new LBMException("Not idle");
      }
      this.cb = cb;
      this.cbArg = cbArg;
      id++;
      rtnId = id;
      tmrState = States.TIMING;
      if (umTimerRecycler != null) {
        // Re-use previous timer object.
        umTimer = umTimerRecycler;
        umTimerRecycler = null;
        umTimer.reschedule(delay);
      } else {
        umTimer = ctx.scheduleTimer(delay, expireCb, null);
      }
    }  // synchronized

    return rtnId;
  }

  // Normal timer expiration.
  public class TmrExpireCallback implements LBMTimerCallback {
    Tmr tmr;
    LBMContext ctx;

    // THREAD: user
    TmrExpireCallback(Tmr tmr, LBMContext ctx) {
      this.tmr = tmr;
      this.ctx = ctx;
    }

    // THREAD: ctx
    public void onExpiration(Object cbArg) {
      try {
        tmr.expire();
      } catch (Exception e) {
        LBMPubLog.pubLog(LBM.LOG_WARNING, "Tmr::TmrExpireCallback::onExpire: tmr.expire threw '" + e.toString() + "'\n");
      }
    }
  }

  // Internal method which invokes user's expiration callback (if appropriate).
  // THREAD: ctx
  private void expire() {
    synchronized (this) {
      umTimerRecycler = umTimer;
      umTimer = null;
      if (tmrState == States.TIMING) {
        tmrState = States.IDLE;
        TmrCallback localCb = cb;
        Object localCbArg = cbArg;
        // Release any resources used by the timer.
        cb = null;
        cbArg = null;
        localCb.onExpire(this, ctx, localCbArg);
      }
    }
  }

  // Public API to cancel a timer, NOT from context thread.
  // On return, guarantees that user callback either has been called or won't be called.
  // THREAD: user
  void cancelSync() throws Exception {
    boolean mustWait = false;
    synchronized (this) {
      if (tmrState == States.CANCEL_PENDING) {
        throw new LBMException("Cancel already pending");
      }
      else if (tmrState == States.TIMING) {
        if (umTimer != null) {
          tmrState = States.CANCEL_PENDING;
          if (umCancelerRecycler != null) {
            // Re-use previous timer object.
            umCanceler = umCancelerRecycler;
            umCancelerRecycler = null;
            umCanceler.reschedule(0);
          } else {
            umCanceler = ctx.scheduleTimer(0, cancelerCb, null);
          }
          mustWait = true;
        }
      }
    }  // synchronized

    // Wait for completion.
    if (mustWait) {
      wakeupSem.acquire();
    }
  }

  // Internal method to process a cancel.
  class TmrCancelerCallback implements LBMTimerCallback {
    Tmr tmr;
    LBMContext ctx;

    // THREAD: user
    TmrCancelerCallback(Tmr tmr, LBMContext ctx) {
      this.tmr = tmr;
      this.ctx = ctx;
    }

    // THREAD: ctx
    public void onExpiration(Object cbArg) {
      try {
        tmr.canceler();
      } catch (Exception e) {
        LBMPubLog.pubLog(LBM.LOG_WARNING, "Tmr::TmrCancelCallback::onExpiration: tmr.canceler threw '" + e.toString() + "'\n");
      }
    }
  }

  // THREAD: ctx
  private void canceler() throws Exception {
    synchronized (this) {
      if (tmrState == States.CANCEL_PENDING) {
        if (umTimer != null) {
          umTimer.cancel();
          umTimerRecycler = umTimer;
          umTimer = null;
        }
        tmrState = States.IDLE;

        // Release any resoureces used by the timer.
        cb = null;
        cbArg = null;

        wakeupSem.release();
      }
    }
  }

  // Public API to cancel timer from context thread.
  // Since already in context thread, don't need to set zero-duration timer.
  // THREAD: ctx
  void cancelCtxThread() throws Exception {
    synchronized (this) {
      if (tmrState == States.CANCEL_PENDING) {
        throw new LBMException("Cancel already pending");
      }
      else if (tmrState == States.TIMING) {
        if (umTimer != null) {
          umTimer.cancel();
          umTimerRecycler = umTimer;
          umTimer = null;
          // Release any resoureces used by the timer.
          cb = null;
          cbArg = null;
        }
      }
      tmrState = States.IDLE;
    }  // synchronized
  }
}

File: oldc/tmr.c
/* tmr.c - better timers built on UM timers.
 *
 * See https://github.com/UltraMessaging/lbmct
 *
 * Copyright (c) 2005-2018 Informatica Corporation. All Rights Reserved.
 * Permission is granted to licensees to use or alter this software for
 * any purpose, including commercial applications, according to the terms
 * laid out in the Software License Agreement.
 *
 * This source code example is provided by Informatica for educational
 * and evaluation purposes only.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND INFORMATICA DISCLAIMS ALL WARRANTIES
 * EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION, ANY IMPLIED WARRANTIES OF
 * NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
 * INFORMATICA DOES NOT WARRANT THAT USE OF THE SOFTWARE WILL BE UNINTERRUPTED
 * OR ERROR-FREE.  INFORMATICA SHALL NOT, UNDER ANY CIRCUMSTANCES, BE LIABLE TO
 * LICENSEE FOR LOST PROFITS, CONSEQUENTIAL, INCIDENTAL, SPECIAL OR INDIRECT
 * DAMAGES ARISING OUT OF OR RELATED TO THIS AGREEMENT OR THE TRANSACTIONS
 * CONTEMPLATED HEREUNDER, EVEN IF INFORMATICA HAS BEEN APPRISED OF THE
 * LIKELIHOOD OF SUCH DAMAGES.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <limits.h>
#ifdef _WIN32
    #include <winsock2.h>
#else
    #include <unistd.h>
    #include <netinet/in.h>
    #include <arpa/inet.h>
#endif
#include <lbm/lbm.h>
#include "tmr.h"
#include "prt.h"  /* Some portability definitions. */


/*
 * Internal functions (not public).
 */


int tmr_expire_cb(lbm_context_t *ctx, const void *clientd)
{
  tmr_t *tmr = (tmr_t *)clientd;

  PRT_MUTEX_LOCK(tmr->lock);

  if (tmr->state == TMR_STATE_TIMING) {
    tmr->um_timer_id = -1;
    /* Set to idle before the callback so that callback can reschedule. */
    tmr->state = TMR_STATE_IDLE;

    /* Call app callback. */
    (*tmr->app_proc)(tmr, tmr->ctx, tmr->app_clientd);
  }

  PRT_MUTEX_UNLOCK(tmr->lock);

  return LBM_OK;
}  /* tmr_expire_cb */


int tmr_cancel_sync_cb(lbm_context_t *ctx, const void *clientd)
{
  tmr_t *tmr = (tmr_t *)clientd;

  PRT_MUTEX_LOCK(tmr->lock);

  if (tmr->state == TMR_STATE_CANCEL_PENDING) {
    if (tmr->um_timer_id != -1) {
      lbm_cancel_timer(tmr->ctx, tmr->um_timer_id, NULL);
      tmr->um_timer_id = -1;
    }
    tmr->state = TMR_STATE_IDLE;
    PRT_SEM_POST(tmr->sync_sem);  /* Wake up the cancel API. */
  }
  else {
    lbm_logf(LBM_LOG_ERR, "tmr_cancel_sync_cb error state=%d [%s:%d]\n",
      (int)tmr->state, BASENAME(__FILE__), __LINE__);
  }

  PRT_MUTEX_UNLOCK(tmr->lock);

  return LBM_OK;
}  /* tmr_cancel_sync_cb */


/*
 * Public APIs
 */


int tmr_create(tmr_t **tmrp, lbm_context_t *ctx)
{
  tmr_t *tmr;

  PRT_MALLOC_N(tmr, tmr_t, sizeof(tmr_t));

  PRT_MUTEX_INIT_RECURSIVE(tmr->lock);
  PRT_SEM_INIT(tmr->sync_sem, 0);  /* Init count to 0. */
  tmr->ctx = ctx;
  tmr->state = TMR_STATE_IDLE;
  tmr->app_proc = NULL;
  tmr->app_clientd = NULL;
  tmr->delay = 0;
  tmr->id = 0;
  tmr->um_timer_id = -1;

  *tmrp = tmr;
  return LBM_OK;
}  /* tmr_create */


int tmr_schedule(int *id, tmr_t *tmr, tmr_cb_proc app_proc, void *clientd,
  lbm_ulong_t delay)
{
  PRT_MUTEX_LOCK(tmr->lock);

  if (tmr->state != TMR_STATE_IDLE) {
    lbm_seterrf(LBM_EINVAL, "Error at %s:%d, 'tmr_schedule error, state=%d`'",
      BASENAME(__FILE__), __LINE__, tmr->state);
    PRT_MUTEX_UNLOCK(tmr->lock);
    return -1;
  }

  tmr->app_proc = app_proc;
  tmr->app_clientd = clientd;
  tmr->delay = delay;
  tmr->state = TMR_STATE_TIMING;
  if (tmr->id == INT_MAX) {
    tmr->id = 1;
  } else {
    tmr->id ++;
  }
  if (id != NULL) {
    *id = tmr->id;
  }

  tmr->um_timer_id = lbm_schedule_timer(tmr->ctx, tmr_expire_cb,
    tmr, NULL, delay);  /* Pass the tmr object as the clientd. */

  PRT_MUTEX_UNLOCK(tmr->lock);
  return LBM_OK;
}  /* tmr_schedule */


int tmr_cancel_sync(tmr_t *tmr)
{
  PRT_MUTEX_LOCK(tmr->lock);

  if (tmr->state == TMR_STATE_CANCEL_PENDING) {
    /* State is cancel pending only briefly; only way this can happen is
     * two threads calling cancel.
     */
    lbm_seterrf(LBM_EINVAL, "Error at %s:%d, 'tmr_cancel_sync error, state=%d`'",
      BASENAME(__FILE__), __LINE__, tmr->state);
    PRT_MUTEX_UNLOCK(tmr->lock);
    return -1;
  }

  /* If UM timer is active, set zero-length timer to cancel it. */
  if (tmr->um_timer_id != -1) {
    tmr->state = TMR_STATE_CANCEL_PENDING;

    /* Schedule zero-duration timer and wait for it to complete. */
    (void)lbm_schedule_timer(tmr->ctx, tmr_cancel_sync_cb,
      tmr, NULL, 0);  /* Pass the tmr object as the clientd. */

    PRT_MUTEX_UNLOCK(tmr->lock);

    PRT_SEM_WAIT(tmr->sync_sem);
  }
  else {
    /* UM timer not active, do nothing. */
    PRT_MUTEX_UNLOCK(tmr->lock);
  }

  return LBM_OK;
}  /* tmr_cancel_sync */


/* Only call from context thread. */
int tmr_cancel_ctx_thread(tmr_t *tmr)
{
  PRT_MUTEX_LOCK(tmr->lock);

  if (tmr->state == TMR_STATE_CANCEL_PENDING) {
    lbm_seterrf(LBM_EINVAL, "Error at %s:%d, 'tmr_cancel_ctx_thread error, state=%d`'",
      BASENAME(__FILE__), __LINE__, tmr->state);
    PRT_MUTEX_UNLOCK(tmr->lock);
    return -1;
  }

  if (tmr->um_timer_id != -1) {
    lbm_cancel_timer(tmr->ctx, tmr->um_timer_id, NULL);
    tmr->um_timer_id = -1;
  }
  tmr->state = TMR_STATE_IDLE;

  PRT_MUTEX_UNLOCK(tmr->lock);

  return LBM_OK;
}  /* tmr_cancel_ctx_thread */


int tmr_delete_sync(tmr_t *tmr)
{
  PRT_MUTEX_LOCK(tmr->lock);

  if (tmr->state == TMR_STATE_TIMING) {
    tmr->state = TMR_STATE_CANCEL_PENDING;

    /* Schedule zero-duration timer and wait for it to complete. */
    (void)lbm_schedule_timer(tmr->ctx, tmr_cancel_sync_cb,
      tmr, NULL, 0);  /* Pass the tmr object as the clientd. */

    PRT_MUTEX_UNLOCK(tmr->lock);

    PRT_SEM_WAIT(tmr->sync_sem);
  }
  else {
    PRT_MUTEX_UNLOCK(tmr->lock);
  }

  PRT_MUTEX_DELETE(tmr->lock);
  PRT_SEM_DELETE(tmr->sync_sem);
  free(tmr);

  return LBM_OK;
}  /* tmr_delete_sync */
